### ISSUE TITLE
DTLS

### ISSUE THREAD

**Issue by jvermillard** [Mainflux/coreflux#1]
Is there an example on how to use it with DTLS (if supported?)
Thanks

**Comment by drasko** [comment]
@jvermillard nope, not in the repo :(. There were only preliminary POC tries using this module: https://github.com/Rantanen/node-dtls, but it is spread over some private forks, never pulled back to the codebase.

Anyway, Coreflux did not advance much because of two reasons:
1. We were more focused on HTTP (for apps) and MQTT in the beginning
2. I have spotted that I'm actually duplicating code with Telefonica (https://github.com/telefonicaid/lwm2m-node-lib), and I do not like duplicating the efforts...

Now, I on the first look there were some things that I did not like with their approach:
- They were focused on [Fiware](https://www.fiware.org/) - and we did not take this path in the beginning
- GNU Affero license (I liked more Apache-2.0)

However after more analysis we definitely decided to migrate whole Mainflux architecture to Fiware, and we are now building an equivalent to what we have today, just using Fiware GEs (http://catalogue.fiware.org/enablers). Lately Telefonica did really a lot of stuff very close to what we were doing, so we started contributing to their codebase (which is actually contributed to Fiware).

So, I suggest you that you look first mentioned [lwm2m-node-lib](https://github.com/telefonicaid/lwm2m-node-lib), and also here: https://github.com/telefonicaid/lightweightm2m-iotagent. If there is some code to be added, I will send it over there via PRs (I am currently doing this over here: https://github.com/telefonicaid/iotagent-mqtt, but next on the list is lwm2m).

In a next few days I will add LWM2M docker similar to this one: https://github.com/Mainflux/mainflux-fiware/blob/master/docker-compose.yml#L40-L49 in our composition and push the latest images on the DockerHub (WIP, here: https://hub.docker.com/u/mainflux/), so it will be easy to test the system...

BTW, I learned most of the things I know about LWM2M from your presentations and Leshan ;), so please do not hesitate to give me any proposition on how this should be organized... For now I am thinking that keeping a copy of what we have so far in the renamed form of `mainflux-lite` and developing it further as a background task (if time left) would be a good idea, in case where leaner and simpler (and Apache-2.0) approach is needed, but I really think that uniting the force on Fiware project would be better for now, especially that we are on the same track with Telefonica guys in the choice of technologies and dev strategies...

**Comment by jvermillard** [comment]
Thanks for the details!
I suppose I need to wait for a resolution of the DTLS issue in the nodejs world.

**Comment by drasko** [comment]
Yeap. We are following closely what is happening here: https://github.com/nodejs/node/issues/2398 and I am hoping that the fact that WebRTC is using it will push the things in both NodeJS and Golang worlds.

**Comment by jvermillard** [comment]
BTW we picked Java over Golang for Leshan because of scandium and bouncy-castle DTLS implementation availability (we use Golang for MQTT).

**Comment by drasko** [comment]
I see... I must admit that Java is main thing that pushed us away from Leshan :(.  I think that Go based approach is the best, but there were still many packages missing (we have at least 4 servers to run!) so we went with NodeJS as a second best (speed, availability, easy deployments, tons of packages). Plus we need to develop some GUIs for Mainflux and AngularJS is a good fit there, so staying within JS is OK for Mainflux as a system for now. However, concerning LWM2M server I would definitively go Golang path. Something like this, I guess: https://github.com/zubairhamed/betwixt

Speaking of MQTT, this is the best I have found: https://github.com/surgemq/surgemq, but it looks like it is not actively maintained...

**Comment by jvermillard** [comment]
Leshan being Java, is just a jar to run: install Java and run it as "java -jar leshan.jar", I did my best to keep it easy to deploy by excluding JEE and other crap.
Now security, bootstrap, block-wise transfer and CoAP retransmit are mandatory for us, so today only wakaama & leshan are working for us. Also I easily trade my language preferences for ready to use code :)

**Comment by drasko** [comment]
Oh, I'm sure that Java implementation is very good choice for professional telecom deployments. It's just that we started building Mainflux to run several servers in parallel (HTTP, WS, MQTT, CoAP) in one application (for the simplicity of implementation) - so what we were looking for were libraries/modules that we could just include in our NodeJS app. This is where Matteo Colina's Mosca MQTT and CoAP servers fit nicely, as they can be used as a module.

But finally, as we are breaking our server into micro-services that will be running in their separate Docker containers, I guess that language choice really does not matter - as long as we can understand maintain the implementation...
